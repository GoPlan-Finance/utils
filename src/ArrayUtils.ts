// noinspection JSUnusedGlobalSymbols

import { ObjectPathUtils } from './ObjectPathUtils';

type groupByFn<T, K> = (value: T, index: number) => K;

// TS 4.4.0 +
//type groupByResult<T, K extends string | number> = { [key : string | number] : T[] }
type groupByResult<T, K extends string | number> = { [key: string]: T[] };

export class ArrayUtils {
  static randomElement<T>(items: T[]): T {
    if (!items.length) {
      throw new Error('provided array is empty');
    }

    return items[Math.floor(Math.random() * (items.length - 1))];
  }

  static fillWith<T>(length: number, initializer: (index: number) => T): T[] {
    return Array.from({ length }, (_, i) => initializer(i));
  }

  static intersect<T>(array1: T[], array2: T[]): T[] {
    return array1.filter(value => array2.includes(value));
  }

  static shuffle<T>(array: T[]): T[] {
    let currentIndex = array.length;

    // While there remain elements to shuffle...
    while (0 !== currentIndex) {
      // Pick a remaining element...
      const randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex -= 1;

      // And swap it with the current element.
      const temporaryValue = array[currentIndex];
      array[currentIndex] = array[randomIndex];
      array[randomIndex] = temporaryValue;
    }

    return array;
  }

  public static groupBy<T, K extends string | number>(
    array: T[],
    keyCb: groupByFn<T, K>
  ): groupByResult<T, K> {
    return array.reduce((result: groupByResult<T, K>, currentValue, index) => {
      const key: K = keyCb(currentValue as T, index);

      result[key] = result[key] || [];

      result[key].push(currentValue);
      return result;
    }, {});
  }

  public static batches<T>(array: T[], perChunk: number): T[][] {
    return Object.values(
      ArrayUtils.groupBy<T, string>(array, (value, index) => {
        return Math.floor(index / perChunk).toString();
      })
    );
  }

  public static unique<T>(arr: T[]): T[] {
    return [...new Set(arr)];
  }

  public static uniqueByKey<T, V = string>(arr: T[], key: keyof T): T[] {
    const out = [];
    const keys: V[] = [];

    for (const item of arr) {
      const kVal = item[key] as unknown as V;

      if (!keys.includes(kVal)) {
        out.push(item);
        keys.push(kVal);
      }
    }

    return out;
  }

  public static sortByKey<T>(arr: T[], propPath: string, ascending = true): T[] {
    const order = ascending ? 1 : -1;

    return arr.sort((a, b) => {
      return ObjectPathUtils.getPathValue(a, propPath) > ObjectPathUtils.getPathValue(b, propPath)
        ? order * 1
        : order * -1;
    });
  }

  /**
   * Find an array element using filterFn, and move it to finalPosition index.
   * @param original
   * @param finalPosition
   * @param filterFn
   */
  static moveIndex<T>(original: T[], finalPosition: number, filterFn: (a: T) => boolean): T[] {
    const index = original.findIndex(filterFn);

    if (index === -1) {
      return original;
    }

    const tmp = original.splice(index, 1);

    original.splice(finalPosition, 0, ...tmp);

    return original;
  }

  static toKeyValueArray<T, V>(
    objects: T[],
    keyName: keyof T | ((elem: T) => string | number),
    valueKey: keyof T | ((elem: T) => V)
  ): { [key: string]: V } {
    const out: { [key: string]: V } = {};

    for (const object of objects) {
      // noinspection UnnecessaryLocalVariableJS
      let keyVal: string | number = null;
      if (typeof keyName === 'function') {
        keyVal = keyName(object);
      } else {
        keyVal = object[keyName] as unknown as string | number;
      }

      // noinspection UnnecessaryLocalVariableJS
      let value = null;
      if (typeof valueKey === 'function') {
        value = valueKey(object);
      } else {
        value = object[valueKey] as unknown as V;
      }

      out[keyVal] = value;
    }

    return out;
  }

  static toKeyArray<T>(objects: T[], key: keyof T): { [key: string]: T } {
    const out: { [key: string]: T } = {};

    for (const object of objects) {
      const keyVal = object[key] as unknown as string;

      out[keyVal] = object;
    }

    return out;
  }

  /**
   * Take an array, compute the values with valueFn() and filter the array by theses values, and retrieve the filtered
   * element and its corresponding calculated value
   *
   * @param elems
   * @param filterFn  Receive two values,  return true if  'a' is preferable than 'b'
   * @param valueFn
   *
   * @example
   * const output = ArrayUtils.filterByvalue(
   *                    [ 1, 5 ,10],
   *                    (a, b ) => a > b,
   *                   elem => elem * 10 + 1 )
   *
   *  console.log(output)
   *  { key : "3" , elem : 10,   val : 101 }
   */
  public static filterByValue<T, V>(
    elems: T[],
    filterFn: (a: V, b: V) => boolean,
    valueFn: (elem: T) => V
  ): { key: string; elem: T; value: V } {
    const keys = Object.keys(elems);
    const values = Object.values(elems);

    if (values.length === 0) {
      return null;
    }

    let index = 0;
    let highestVal: V = undefined;

    for (let i = 0; i < values.length; i++) {
      const val = valueFn(values[i]);

      if (highestVal === undefined || filterFn(val, highestVal)) {
        index = i;
        highestVal = val;
      }
    }

    return {
      key: keys[index],
      elem: values[index],
      value: highestVal,
    };
  }

  public static sum<T>(arr: T[], cb: (item: T) => number): number {
    return arr.reduce((result, current) => result + cb(current), 0);
  }
}
