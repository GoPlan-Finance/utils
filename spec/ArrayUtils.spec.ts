import { ArrayUtils } from '@utils/ArrayUtils';

describe('ArrayUtils', () => {
  describe('fillWith', () => {
    it('should call initializer on every items', function () {
      let i = 1;
      const initializer = () => i++;

      const val = ArrayUtils.fillWith(4, initializer);
      expect(val).toStrictEqual([1, 2, 3, 4]);
    });

    it('should remove duplicate items', function () {
      const data = [1, 2, 3, 4, 4, 4];

      const val = ArrayUtils.unique(data);
      expect(val).toStrictEqual([1, 2, 3, 4]);
    });
  });

  describe('unique', () => {
    it('should return all non unique items', function () {
      const data = [1, 2, 3, 4];

      const val = ArrayUtils.unique(data);
      expect(val).toStrictEqual([1, 2, 3, 4]);
    });

    it('should remove duplicate items', function () {
      const data = [1, 2, 3, 4, 4, 4];

      const val = ArrayUtils.unique(data);
      expect(val).toStrictEqual([1, 2, 3, 4]);
    });
  });

  describe('intersect', () => {
    it('should return matching elements in two arrays', function () {
      expect(ArrayUtils.intersect([1, 2, 3], [3, 4, 5, 2])).toStrictEqual([2, 3]);
    });

    it('intersect should return nothing if one of the array is empty', function () {
      expect(ArrayUtils.intersect([1, 2, 3], [])).toStrictEqual([]);
      expect(ArrayUtils.intersect([], [1, 2, 3])).toStrictEqual([]);
    });
  });

  describe('groupBy', () => {
    it('should return matching elements grouped by sub-keys', function () {
      const data = [
        { color: 'blue', n: 1 },
        { color: 'blue', n: 2 },
        { color: 'red', n: 3 },
        { color: 'red', n: 4 },
        { color: 'yellow', n: 5 },
      ];

      const result = ArrayUtils.groupBy(data, item => item.color);
      expect(result).toMatchSnapshot();

      const result2 = ArrayUtils.groupBy(data, item => (item.n % 2 ? 'odd' : 'even'));
      expect(result2).toMatchSnapshot();
    });
  });

  describe('batches', () => {
    it('should return matching elements grouped by batch number', function () {
      const arr = Array.from(Array(100).keys());

      expect(ArrayUtils.batches(arr, 10)).toMatchSnapshot();

      expect(ArrayUtils.batches(arr, 200)).toMatchSnapshot();

      expect(ArrayUtils.batches([], 200)).toMatchSnapshot();
    });

    it('should return a single batch when perChunk is bigger than len', function () {
      const arr = Array.from(Array(100).keys());
      expect(ArrayUtils.batches(arr, 200)).toMatchSnapshot();
    });

    it('should return an empty array when no data is given', function () {
      expect(ArrayUtils.batches([], 200)).toMatchSnapshot();
    });
  });

  describe('randomElement', () => {
    it('should return one of the array element randomly', () => {
      const setMock = (val: number) => {
        const mockMath = Object.create(global.Math);
        mockMath.random = () => val;
        global.Math = mockMath;
      };

      const arr = Array.from(Array(100).keys());

      setMock(0);
      expect(ArrayUtils.randomElement(arr)).toBe(0);

      setMock(0.01);
      expect(ArrayUtils.randomElement(arr)).toBe(0);

      setMock(0.5);
      expect(ArrayUtils.randomElement(arr)).toBe(49);

      setMock(1);
      expect(ArrayUtils.randomElement(arr)).toBe(99);
    });

    it('should return one of the array element randomly', () => {
      const setMock = (val: number) => {
        const mockMath = Object.create(global.Math);
        mockMath.random = () => val;
        global.Math = mockMath;
      };

      const arr = Array.from(Array(100).keys());

      setMock(0);
      expect(ArrayUtils.randomElement(arr)).toBe(0);

      setMock(0.01);
      expect(ArrayUtils.randomElement(arr)).toBe(0);

      setMock(0.5);
      expect(ArrayUtils.randomElement(arr)).toBe(49);

      setMock(1);
      expect(ArrayUtils.randomElement(arr)).toBe(99);
    });

    it('should throw when provided array is empty', () => {
      const arr = Array.from(Array(100).keys());

      expect(() => ArrayUtils.randomElement([])).toThrowError('provided array is empty');
    });
  });

  describe('filterByvalue', () => {
    it('should return the filter element and value', () => {
      const output = ArrayUtils.filterByValue(
        [1, 5, 10],
        (a, b) => a > b,
        elem => elem * 10 + 1
      );

      expect(output).toStrictEqual({
        key: '2',
        elem: 10,
        value: 101,
      });
    });

    it('should throw when provided array is empty', () => {
      const arr = Array.from(Array(100).keys());

      expect(() => ArrayUtils.randomElement([])).toThrowError('provided array is empty');
    });
  });
});
