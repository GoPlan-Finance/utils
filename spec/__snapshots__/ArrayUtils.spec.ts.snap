// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayUtils batches should return a single batch when perChunk is bigger than len 1`] = `
[
  [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
    30,
    31,
    32,
    33,
    34,
    35,
    36,
    37,
    38,
    39,
    40,
    41,
    42,
    43,
    44,
    45,
    46,
    47,
    48,
    49,
    50,
    51,
    52,
    53,
    54,
    55,
    56,
    57,
    58,
    59,
    60,
    61,
    62,
    63,
    64,
    65,
    66,
    67,
    68,
    69,
    70,
    71,
    72,
    73,
    74,
    75,
    76,
    77,
    78,
    79,
    80,
    81,
    82,
    83,
    84,
    85,
    86,
    87,
    88,
    89,
    90,
    91,
    92,
    93,
    94,
    95,
    96,
    97,
    98,
    99,
  ],
]
`;

exports[`ArrayUtils batches should return an empty array when no data is given 1`] = `[]`;

exports[`ArrayUtils batches should return matching elements grouped by batch number 1`] = `
[
  [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
  ],
  [
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
  ],
  [
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
  ],
  [
    30,
    31,
    32,
    33,
    34,
    35,
    36,
    37,
    38,
    39,
  ],
  [
    40,
    41,
    42,
    43,
    44,
    45,
    46,
    47,
    48,
    49,
  ],
  [
    50,
    51,
    52,
    53,
    54,
    55,
    56,
    57,
    58,
    59,
  ],
  [
    60,
    61,
    62,
    63,
    64,
    65,
    66,
    67,
    68,
    69,
  ],
  [
    70,
    71,
    72,
    73,
    74,
    75,
    76,
    77,
    78,
    79,
  ],
  [
    80,
    81,
    82,
    83,
    84,
    85,
    86,
    87,
    88,
    89,
  ],
  [
    90,
    91,
    92,
    93,
    94,
    95,
    96,
    97,
    98,
    99,
  ],
]
`;

exports[`ArrayUtils batches should return matching elements grouped by batch number 2`] = `
[
  [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
    30,
    31,
    32,
    33,
    34,
    35,
    36,
    37,
    38,
    39,
    40,
    41,
    42,
    43,
    44,
    45,
    46,
    47,
    48,
    49,
    50,
    51,
    52,
    53,
    54,
    55,
    56,
    57,
    58,
    59,
    60,
    61,
    62,
    63,
    64,
    65,
    66,
    67,
    68,
    69,
    70,
    71,
    72,
    73,
    74,
    75,
    76,
    77,
    78,
    79,
    80,
    81,
    82,
    83,
    84,
    85,
    86,
    87,
    88,
    89,
    90,
    91,
    92,
    93,
    94,
    95,
    96,
    97,
    98,
    99,
  ],
]
`;

exports[`ArrayUtils batches should return matching elements grouped by batch number 3`] = `[]`;

exports[`ArrayUtils groupBy should return matching elements grouped by sub-keys 1`] = `
{
  "blue": [
    {
      "color": "blue",
      "n": 1,
    },
    {
      "color": "blue",
      "n": 2,
    },
  ],
  "red": [
    {
      "color": "red",
      "n": 3,
    },
    {
      "color": "red",
      "n": 4,
    },
  ],
  "yellow": [
    {
      "color": "yellow",
      "n": 5,
    },
  ],
}
`;

exports[`ArrayUtils groupBy should return matching elements grouped by sub-keys 2`] = `
{
  "even": [
    {
      "color": "blue",
      "n": 2,
    },
    {
      "color": "red",
      "n": 4,
    },
  ],
  "odd": [
    {
      "color": "blue",
      "n": 1,
    },
    {
      "color": "red",
      "n": 3,
    },
    {
      "color": "yellow",
      "n": 5,
    },
  ],
}
`;

exports[`ArrayUtils toKeyValueArray should return matching elements grouped by keys 1`] = `
{
  "blue": {
    "color": "blue",
    "n": 2,
  },
  "red": {
    "color": "red",
    "n": 4,
  },
  "yellow": {
    "color": "yellow",
    "n": 5,
  },
}
`;

exports[`ArrayUtils toKeyValueArray should return matching elements grouped by keys 2`] = `
{
  "blue": 2,
  "red": 4,
  "yellow": 5,
}
`;
